### DISCUSSION TITLE
Analog voltage reading generates action

### DISCUSSION THREAD

**Discussion by richonguzman** [HomeSpan/HomeSpan#269] [General]
Thanks a lot for this library, just have an ESP32 with a bigpowerbank (18650) working to turn lights (with mosfets as high power switches)!!!

But i would Love to have a way to analogRead() voltage (like a LDR or voltage divider) to show battery charging status of the powerbank and know when to charge it.

Or the LDR to trigger like a button to turn on the lights

What should I look into??

**Comment by HomeSpan** [comment]
You can include any code you'd like in the *loop()* method of a Service, including using some form of analogRead() to read the charging status of your batteries.  Though you'll need to create your methods specific to the battery, once you have the value you can then trigger other actions, as well as use the *setVal()* method to update other Characteristics.  See the [Service Loops Tutorial](https://github.com/HomeSpan/HomeSpan/tree/master/examples/12-ServiceLoops) for details on how to use the *loop()* method.

**Reply by richonguzman** [reply]
thanks, can I post here the .ino and .h which I modified to check if everything is well defined?

**Comment by HomeSpan** [comment]
Sure thing, though the best test is to try it out in practice.  At a minimum you'd know if it compiles correctly.  When testing a new device I often do it in stages, incrementally adding in new Services and Characteristics that return a fixed value I can use to check things are working.  After I know all the HomeSpan stuff is working fine, then I layer in the additional code and logic need to interface with whatever hardware I'm using (in your case, reading an analog voltage representing battery charge).

**Comment by richonguzman** [comment]
here is the .ino

`#include "HomeSpan.h" 
#include "DEV_Identify.h"      
#include "DEV_Battery.h" 

void setup() {
  Serial.begin(115200);
  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");

  new SpanAccessory();  
    new DEV_Identify("Bridge #1","HomeSpan","123-ABC","HS Bridge","0.9",3);
    new Service::HAPProtocolInformation();
      new Characteristic::Version("1.1.0");
      
  new SpanAccessory();                                                          
    new DEV_Identify("Battery","HomeSpan","123-ABC","BatterySensor","0.9",0);
    new DEV_Medidor_Bateria();
}

void loop(){  
  homeSpan.poll();
}`


here the 'Dev_Battery.h'
`struct DEV_Medidor_Bateria : Service::BatteryService { 
  SpanCharacteristic *charge;
  
  DEV_Medidor_Bateria() : Service::BatteryService(){ 
    charge=new Characteristic::BatteryLevel(0);
    charge->setRange(0,100,1);
  } 

  void loop(){
    if(charge->timeVal()>5000){
      uint8_t voltage=charge->getVal<uint8_t>(); 
      
      charge->setVal(voltage);
      
      LOG1("voltage Update: ");
      LOG1(voltage);
      LOG1("\n");            
    }
  }  
};`


So, after a lot of reading , I can't find how to tell the ESP which pin to read the voltage (with a voltage divider

**Comment by HomeSpan** [comment]
The sketch above seems reasonable, but you would not use getVal() in the loop() method, only setVal().

Note there's nothing specific in HomeSpan about reading analog voltages.  You would just write normal Arduino/ESP32 code to read one of the ADCs on your ESP32.  Each ESP32 is different so the pin number you chose will depend on your specific board.  Typical Arduino/ESP32 code to read analog voltage looks something like: ` sensorValue = analogRead(sensorPin);`.  The Arduino web site should have a lot of tutorials and examples on how to do this.  You may want to experiment first using a simple sketch without HomeSpan.  Then, once you can read the voltage and convert the scale to %charge, you can add the logic into a HomeSpan sketch and pass the value of sensorValue back to the Home App using `setVal(sensorValue*scaling_factor);`

**Reply by richonguzman** [reply]
awesome, I already know how to read values on Arduino or ESP32, but wanted to be very cautious about how to define which pin acts as the voltage-reader and will declare It. But should I declare it in the Dev_Battery.h loop, right?

**Reply by Xorlent** [reply]
Gregg, could I ask why you say not to use getVal() in a loop method, or is that recommendation only based on the INO Richon posted?  I've used that quite a bit in my sketches with no adverse effects that I can tell.

**Reply by HomeSpan** [reply]
Sorry, I should have been more specific.  Using getVal() inside loop() is perfectly fine and I use it quite a bit for many different sketches.  My comment was rather about the use case above, in which you aren't trying to retrieve the voltage by reading the value associated with the BatteryLevel Characteristic, but rather you are trying to set the value of that Characteristic based on what the actual pin voltage is.  Setting the value of the Characteristic with setVal() informs the Home App of the new value.  Getting the value of the Characteristic would be applicable if the value itself was being controlled by the user inside the Home App, which is generally not applicable for a BatteryLevel sensor (it is a read-only Characteristic from the perspective of the Home App).

**Reply by HomeSpan** [reply]
Also @richonguzman, yes it's fine to declare the sensor pin inside the Dev_Battery loop.   You can also make it a parameter of Dev_Battery Service itself, in case you ever need to read voltages from more than one pin for multiple battery packs.  This allows you to re-use the code.  See [Example 5](https://github.com/HomeSpan/HomeSpan/tree/master/examples/05-WorkingLED) for how to do this.

**Comment by richonguzman** [comment]
turns out great (with some minor problem)

I can "analogRead()" fine and even do a 50 sample to avoid reading errors, later transform the reading into the real Voltage and then transform into a % of charge which finally was rounded for easy read.

later the fight to show it in HomeApp, which appeared and setVal() worked fine.

sadly the battery tile shows as a HOME icon , shows that is incompatible , although when you open it changes (after the normal 5 seconds) into the new analogRead() value read.

so it works , it updates, but it does not show as a informative and direct tile/icon like a thermometer 


should I do a hard-core modification of a termomether for this?

**Comment by richonguzman** [comment]
more news:

add the to missing characteristic of BatteryService:  ChargingState, StatusLowBattery in my project (3 are required)

now I can set a value of the battery which shows "low battery" although it still doesn't like to be treated as an accessory (it says incompatible)
